File: library/src/main/java/pt/ulisboa/tecnico/aux/Keys.java
package pt.ulisboa.tecnico.aux;

import lombok.Getter;
import lombok.Setter;

import javax.crypto.KeyGenerator;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.ByteArrayOutputStream;
import java.security.Key;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.spec.X509EncodedKeySpec;

import static pt.ulisboa.tecnico.aux.Constants.*;
import static pt.ulisboa.tecnico.aux.Constants.SYM_ALGO;
import static pt.ulisboa.tecnico.aux.FileReader.*;
import static pt.ulisboa.tecnico.aux.Conversion.*;

@Getter
public class Keys {

    @Setter(lombok.AccessLevel.NONE)
    private Key secretKey; // KEK

    private Key secretSessionKey;

    private Key publicKey;

    private Key privateKey;

    private Key receiverPublicKey;

    private final IvParameterSpec iv = new IvParameterSpec(Constants.IV);

    public Keys(String secretKeyPath) throws Exception {
        assignSecretKey(secretKeyPath);
        generateAsymKey();
    }

    private void assignSecretKey(String secretKeyPath) throws Exception {
        byte[] encoded = FileReader.readFile(secretKeyPath);
        secretKey = new SecretKeySpec(encoded, SYM_ALGO);
    }

    private void generateAsymKey() throws Exception {
        KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(ASYM_ALGO);
        keyPairGenerator.initialize(ASYM_KEY_SIZE);
        KeyPair keyPair = keyPairGenerator.generateKeyPair();
        publicKey = keyPair.getPublic();
        privateKey = keyPair.getPrivate();
    }

    public void generateSecretSessionKey() throws Exception {
        KeyGenerator keyGenerator = KeyGenerator.getInstance(SYM_ALGO);
        keyGenerator.init(SYM_KEY_SIZE);
        secretSessionKey = keyGenerator.generateKey();
    }

    public byte[] creationPayload() throws Exception {
        byte[] encodedSecretSessionKey = secretSessionKey.getEncoded();
        byte[] encodedPublicKey = publicKey.getEncoded();
        ByteArrayOutputStream payload = new ByteArrayOutputStream();

        if (!write(intToBytes(encodedSecretSessionKey.length), payload)) throw new Exception(
            "Error writing to byte array");
        if (!write(encodedSecretSessionKey, payload)) throw new Exception("Error writing to byte array");

        if (!write(intToBytes(encodedPublicKey.length), payload)) throw new Exception("Error writing to byte array");
        if (!write(encodedPublicKey, payload)) throw new Exception("Error writing to byte array");

        return payload.toByteArray();
    }

    public byte[] receiveSessionKey(byte[] input) throws Exception {
        int secretSessionKeyLength = bytesToInt(input, 0).orElseThrow(() -> new Exception(
            "Error reading from byte array"));
        byte[] encodedSecretSessionKey = read(input, INT_SIZE, secretSessionKeyLength).orElseThrow(() -> new Exception(
            "Error reading from byte array"));
        secretSessionKey = new SecretKeySpec(encodedSecretSessionKey, SYM_ALGO);

        int publicKeyLength = bytesToInt(input, INT_SIZE + secretSessionKeyLength).orElseThrow(() -> new Exception(
            "Error reading from byte array"));
        byte[] encodedReceiverPublicKey = read(input, INT_SIZE + secretSessionKeyLength + INT_SIZE, publicKeyLength)
            .orElseThrow(() -> new Exception("Error reading from byte array"));
        receiverPublicKey = KeyFactory.getInstance(ASYM_ALGO).generatePublic(new X509EncodedKeySpec(
            encodedReceiverPublicKey));

        byte[] encodedPublicKey = publicKey.getEncoded();
        ByteArrayOutputStream payload = new ByteArrayOutputStream();

        if (!write(intToBytes(encodedPublicKey.length), payload)) throw new Exception("Error writing to byte array");
        if (!write(encodedPublicKey, payload)) throw new Exception("Error writing to byte array");

        return payload.toByteArray();
    }

    public void receivePublicKey(byte[] input) throws Exception {
        int publicKeyLength = bytesToInt(input, 0).orElseThrow(() -> new Exception("Error reading from byte array"));
        byte[] encodedPublicKey = read(input, INT_SIZE, publicKeyLength).orElseThrow(() -> new Exception(
            "Error reading from byte array"));
        receiverPublicKey = KeyFactory.getInstance(ASYM_ALGO).generatePublic(new X509EncodedKeySpec(encodedPublicKey));
    }

}

File: libraryTS/src/app/aux/Keys.ts
import * as crypto from 'crypto';
import * as fs from 'fs';

import { intToBytes, bytesToInt, read, write } from './Conversion';
import { readFile } from './FileReader';
import { Constants } from './Constants';

export class Keys {
    private secretKey: crypto.KeyObject | null = null; // KEK

    private secretSessionKey: crypto.KeyObject | null = null;

    private publicKey: crypto.KeyObject | null = null;

    private privateKey: crypto.KeyObject | null = null;

    private receiverPublicKey: crypto.KeyObject | null = null;

    private readonly iv: Buffer = Buffer.from(Constants.IV);

    constructor(secretKeyPath: string) {
        this.assignSecretKey(secretKeyPath);
    }

    private async assignSecretKey(secretKeyPath: string): Promise<void> {
        const encoded = await readFile(secretKeyPath);
        this.secretKey = crypto.createSecretKey(encoded);
    }

    async generateAsymKey(): Promise<void> {
        const { publicKey, privateKey } = crypto.generateKeyPairSync(Constants.ASYM_ALGO, {
            modulusLength: Constants.ASYM_KEY_SIZE,
            publicKeyEncoding: { type: 'spki', format: 'der' },
            privateKeyEncoding: { type: 'pkcs8', format: 'der' },
        });

        this.publicKey = publicKey;
        this.privateKey = privateKey;
    }

    async generateSecretSessionKey(): Promise<void> {
        this.secretSessionKey = crypto.createSecretKey(crypto.randomBytes(Constants.SYM_KEY_SIZE / 8));
    }

    creationPayload(): Buffer {
        const encodedSecretSessionKey = this.secretSessionKey!.export();
        const encodedPublicKey = this.publicKey!.export({ type: 'spki', format: 'der' });

        const payload = Buffer.concat([
            intToBytes(encodedSecretSessionKey.length),
            encodedSecretSessionKey,
            intToBytes(encodedPublicKey.length),
            encodedPublicKey,
        ]);

        return payload;
    }

    receiveSessionKey(input: Buffer): void {
        const secretSessionKeyLength = bytesToInt(input, 0)!;
        const encodedSecretSessionKey = read(input, Constants.INT_SIZE, secretSessionKeyLength)!;
        this.secretSessionKey = crypto.createSecretKey(encodedSecretSessionKey);

        const publicKeyLength = bytesToInt(
            input,
            Constants.INT_SIZE + secretSessionKeyLength
        )!;
        const encodedReceiverPublicKey = read(
            input,
            Constants.INT_SIZE + secretSessionKeyLength + Constants.INT_SIZE,
            publicKeyLength
        )!;
        this.receiverPublicKey = crypto.createPublicKey({
            key: encodedReceiverPublicKey,
            format: 'der',
            type: 'spki',
        });
    }

    receivePublicKey(input: Buffer): void {
        const publicKeyLength = bytesToInt(input, 0)!;
        const encodedPublicKey = read(input, Constants.INT_SIZE, publicKeyLength)!;
        this.receiverPublicKey = crypto.createPublicKey({
            key: encodedPublicKey,
            format: 'der',
            type: 'spki',
        });
    }

    getPublicKeyPayload(): Buffer {
        const encodedPublicKey = this.publicKey!.export({ type: 'spki', format: 'der' });
        const payload = Buffer.concat([intToBytes(encodedPublicKey.length), encodedPublicKey]);
        return payload;
    }

    clearKeys(): void {
        this.privateKey = null;
        this.publicKey = null;
        this.secretSessionKey = null;
        this.receiverPublicKey = null;
    }
}
